I made the behavior of NewDefaultClient and Dial configurable

eyedeekay opened the pull request:
In situations where the SAM bridge isn't on the local host, the existing Dial function won't work because it always creates the Client using NewDefaultClient(which always points to localhost:7656). I decided to change this behavior in my branch to allow the behavor of NewDefaultClient to be configured using 2 variables, SamHostAddress and SamHostPort. I think this is an improvement to the way the Dial function works. The default values of SamHostAddress and SamHostPort are 127.0.0.1 and 7656 respectively, so this change shouldn't affect the behavior of applications unless they explicitly set the new variables.

eyedeekay committed:
I don't know exactly why, but checking that the kvPair in replyParser.go is not nil fixes the crash on image-heavy sites.

Configuration variables to set sam bridge address and port

eyedeekay committed:
Configuration variables to set sam bridge address and port

eyedeekay committed:
Configuration variables to set sam bridge address and port

eyedeekay committed:
Configuration variables to set sam bridge address and port

eyedeekay committed:
configurable default SAM bridge

remove excessive file

eyedeekay committed:
configurable default SAM bridge

eyedeekay committed:
remove excessive file

cryptix commented:
I'm in favor of the option to do this but don't like having global variables as configuration. With this approach you can't created multiple clients with different configurations without being careful about temporal ordering during their creation...

In this case `samHost` and `samPort` would more cleanly fit into the `Client struct{}`.

See my comment on issue #5 for a better design.

eyedeekay committed:
Re-doing as functional options instead. NewClient is backward-compatible.

eyedeekay committed:
Re-doing as functional options instead. NewClient is backward-compatible.

eyedeekay committed:
Re-doing as functional options instead. NewClient is backward-compatible.

eyedeekay committed:
Re-doing as functional options instead. NewClient is backward-compatible.

eyedeekay committed:
wrote some tests.

eyedeekay committed:
wrote some tests.

forgot to gofmt

removed reference to setup from options_test.go

reversed relative import in example

eyedeekay commented:
I made the suggested changes and turned the old NewClient() into a backward-compatible wrapper for NewClientFromOptions, and also wrote some tests for them.

eyedeekay committed:
forgot to gofmt

eyedeekay committed:
removed reference to setup from options_test.go

cryptix commented on options.go:
```diff
@@ -0,0 +1,29 @@
+		if err != nil {
+			return err
+		}
+		if port < 65536 && port > -1 {
```
hmm.. I'd maybe use this as an error case and return _invalid port_.

eyedeekay replied:
I am ready to make this change, but should the error indicate that the reason for the invalid port is the conversion from string to integer? Like, "Invalid port; non-number."

cryptix replied:
oh sorry. i misread this.

i glanced over the negative `if` because i'm more used to code like this:

```go
if os.IsEOF(err) {
  // cleanup or log
  return err
}
// go on
```

than 

```go
if _what i want_ {
  // go on
} else {
   // cleanup or log
  return err
}

```

cryptix replied:
In the upper case i prefer to use pkg/errors to wrap it like `errors.Wrap(err, "port: invalid input port")` then you also see the strconv reason with context.

cryptix commented on options_test.go:
```diff
@@ -0,0 +1,29 @@
+}
+
+func TestOptionsHello(t *testing.T) {
+	ConnDebug = true
```
`ConnDebug` could also be an option, maybe?

cryptix commented on options_test.go:
```diff
@@ -0,0 +1,29 @@
+func TestOptionsHello(t *testing.T) {
+	ConnDebug = true
+	optaddr(t)
+	optport(t)
```
I'd just make two normal tests `TestOptionAddr` and `TestOptionPort`.

cryptix commented on example/httpTest.go:
```diff
@@ -6,8 +6,7 @@ import (
 
-	//"github.com/cryptix/goSam"
-    ".."
+	".."
```
uuuh... relative imports are a no-no in my book.

eyedeekay replied:
My bad. I did that to make sure it still worked after my changes and forgot to change it back.

cryptix reviewed (commented):
some nitpicks but looks really good already!

eyedeekay committed:
reversed relative import in example

eyedeekay committed:
fixed the tests and the error case in set port

eyedeekay committed:
fixed the tests and the error case in set port

made debug an option rather than a global

eyedeekay committed:
made debug an option rather than a global

eyedeekay committed:
made debug an option rather than a global

Those aren't supposed to be in there yet.

Those aren't supposed to be in there yet.

eyedeekay committed:
made debug an option rather than a global

eyedeekay committed:
Those aren't supposed to be in there yet.

eyedeekay committed:
Those aren't supposed to be in there yet.

cryptix commented on options.go:
```diff
@@ -0,0 +1,37 @@
+
+type Option func(*Client) error
+
+func SetAddr(s string) func(*Client) error {
```
The pedantic me would say the names here are a bit wonky.

* `SetAddr()` should take a "host:port" string
* add another `SetHost(string)`

eyedeekay replied:
I made it so that SetAddr could take "host:port" "host" port(int), or "host" "port" if that's OK. If it's trying to do too much I can just change it so that it takes one string, splits it, and sets the variables. Each possible combination has been tested though.

Also, I'm new to the functional options thing(I taught myself go from internet resources, my learning is biased toward what I know to look for), and wasn't sure if I could call SetHost and SetPort from within SetAddr. I'm still unclear on that? If so I won't have to repeat the validation stuff in the new SetAddr.

cryptix commented on client.go:
```diff
@@ -24,18 +27,42 @@ func NewDefaultClient() (*Client, error) {
+
+//return the combined addr:port of the SAM bridge
+func (c *Client) samaddr() string {
+	return c.addr + ":" + c.port
```
i personally practice `fmt.Sprintf("%s:%s", c.addr,c.port)` but not that big of a difference here.

cryptix reviewed (changes requested):
Hope I'm not too annoying here.. after these I think we can merge this.

Also see my two comments on the outdated diff: https://github.com/cryptix/goSam/pull/4#discussion_r186192373

I think you need to rebase #6 afterwards

eyedeekay committed:
Changed SetAddr to take host:port, host, port, port can be either a string or an int.

eyedeekay committed:
Changed SetAddr to take host:port, host, port, port can be either a string or an int.

eyedeekay committed:
test all cases of SetAddr

eyedeekay committed:
test all cases of SetAddr

eyedeekay committed:
use fmt.Sprintf

eyedeekay committed:
use fmt.Sprintf

eyedeekay committed:
SetPort now accepts either a string or an integer port.

eyedeekay committed:
SetPort now accepts either a string or an integer port.

eyedeekay committed:
Merge branch 'master' of github.com:eyedeekay/gosam

eyedeekay commented:
I have completely screwed up the master branch history in my fork trying to rebase. I'm going to close this until I fix it.
